docs/adr/2023-08-18_025-resource-based-api.md

```diff
@@ -0,0 +1,146 @@
+---
+slug: 25
+title: |
+  25. Event-sourced, resource-based API
+authors: []
+tags: [Draft]
+---
+
+## Status
+
+Draft
+
+## Context
+
+- [ADR-3](/adr/3) concluded that a full-duplex communication channels are
+  desirable to interact with a _reactive_ system.
+
+- The Client API communicates several types of messages to clients. Currently
+  this ranges from node-level `PeerConnected`, over head-specific `HeadIsOpen`
+  to messages about transactions like `TxValid`. These messages are all of type
+  `ServerOutput`.
+
+- Current capabilities of the API:
+
+  - Clients can retrieve the whole history of `ServerOutput` messages or
+    opt-out using a query parameter - all or nothing.
+
+  - There is a welcome message called `Greetings` which is always sent, that
+    contains the last `headStatus`.
+
+  - There exists a `GetUTxO` query-like `ClientInput`, which will respond with a
+    `GetUTxOResponse` containing the confirmed UTxO in an open head, or (!) the
+    currently committed UTxO when the head is initializing.
+
+  - While overall `json` encoded, clients can choose choose between `json` or
+    binary (`cbor`) output of `transaction` fields in several of these using a
+    query parameter.
+
+- Many of these features have been added in a "quick and dirty" way, by monkey
+  patching the encoded JSON.
+
+- The current capabalities even do not satisfy all user needs:
+
+  - Need to wade through lots of events to know the latest state (except the
+    very basic `headStatus` from the `Greetings`).
+
+  - Need to poll `GetUTxO` _or_ aggregate confirmed transactions on client side
+    to know the latest UTxO for constructing transactions.
+
+  - Inclusion of the whole UTxO in the head is not always desirable and
+    filtering by address would be beneficial. (not addressed in this ADR though)
+
+  - As [ADR-15](/adr/15) also proposes, some clients may not need (or should
+    not have) access to administrative information.
+
+- It is often a good idea to separate the responsibilities of Commands and
+  Queries (CQRS), as well as the model they use.
+
+## Decision
+
+- Do not add a general purpose querying interface (e.g. graphql) to the
```

Comment:
I assume you don't want to add any persistence a part from the one we have right now (dump to file) so how are these query models suppose to work? Would you add different file for different models? If not how would you get the latest known state for each model? Seems like you would need to search the file to get the output you are looking for or perhaps I am missing something?

Reply:
Ok, I see. You want to use projections to do this.

Reply:
It feels a bit weird this negative statement comes first in the `Decision` section. Perhaps this should be rather added to the `Consequences` section?

Reply:
I just realize.. what this ADR evolved into, is exactly what graphql could provide. The [graphql subscriptions](https://www.apollographql.com/docs/react/data/subscriptions/) are non-standard though, but work basically the same as described in this ADR

Reply:
I removed this statement.